src/platform/testing/e2e/form-tester/form-filler.js

```diff
@@ -0,0 +1,266 @@
+    log(
+      'Field list:',
+      await page.$$eval(FIELD_SELECTOR, elements =>
+        elements.map(e => e.name || e.id),
+      ),
+    );
+    const fields = (await page.$$eval(FIELD_SELECTOR, elements => {
+      // This whole function is executed in the browser and can't contain references
+      //  to anything outside of the local scope.
+      const selectors = new Set();
+      return elements.map(element => {
+        let type = element.type || element.tagName;
+        let selector = element.name || element.id;
+
+        const isDateField = sel =>
+          sel.endsWith('Year') || sel.endsWith('Month') || sel.endsWith('Day');
+        if (isDateField(selector)) {
+          type = 'date';
+          // We only have one date field in the test data, so we fill two or three
+          //  fields with one enterData call
+          selector = selector.replace(/(Year|Month|Day)$/, '');
+        }
+
+        // Make sure not to duplicate entries
+        //  (Specifically useful for radio buttons and date fields)
+        if (selectors.has(selector)) {
+          return null;
+        }
+
+        // Add the item to the set for easy lookup
+        selectors.add(selector);
+
+        return {
+          type,
+          selector,
+        };
+      });
+    }))
+      .filter(field => field) // Duplicates are null fields
+      .filter(field => !touchedFields.has(field.selector))
+      .filter(field => field.selector.startsWith('root_')); // Only grab form fields
+
+    for (const field of fields) {
+      touchedFields.add(field.selector);
+      await enterData(page, field, findData(field.selector, testData), log);
+    }
+
+    // TODO: Check for arrays and click the appropriate add buttons
+    // TODO: Check for multiple add buttons and warn if there are
+
+    newFieldCount = await page.$$eval(
+      FIELD_SELECTOR,
+      elements => elements.length,
+    );
+  } while (fieldCount !== newFieldCount);
+  /* eslint-enable no-await-in-loop */
+};
+
+/**
+ * This is the main entry point. After all the setup has been performed, this function
+ *  loops through the pages, filling in all the data it can until it gets to the review
+ *  page.
+ */
+const fillForm = async (page, testData, testConfig, log) => {
+  // We want these actions to be performed synchronously, so the await
+  //  statements in the loop make sense.
+  /* eslint-disable no-await-in-loop */
+  while (!page.url().endsWith('review-and-submit')) {
+    log(page.url());
+    // TODO: Run axe checker
+
+    // If there's a page hook, run that
+    const url = page.url();
+    const hook = _.get(`pageHooks.${parseUrl(url).path}`, testConfig);
+    if (hook) {
+      if (typeof hook !== 'function') {
+        throw new Error(
+          `Bad testConfig: Page hook for ${url} is not a function`,
+        );
+      }
+      const retVal = hook(page, testData, testConfig);
+      if (retVal instanceof Promise) {
+        await retVal;
+      }
```

Comment:
The nested `if`s can get a little confusing when a nested condition ends with an `if`, but not sure what the alternative is.